iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+  static bool
+  equal(Key lhs, Key rhs)
+  {
+    return lhs == rhs;
+  }
+};
+
+// The RefCountCachePartition is simply a map of key -> Ptr<YourClass>
+// We partition the cache to reduce lock contention
+template <class C> class RefCountCachePartition
+{
+public:
+  RefCountCachePartition(unsigned int part_num, uint64_t max_size, unsigned int max_items, RecRawStatBlock *rsb = NULL);
+  Ptr<C> get(uint64_t key);
+  void put(uint64_t key, C *item, int size = 0, int expire_time = 0);
+  void erase(uint64_t key, int expiry_time = -1);
+
+  void clear();
+  bool is_full();
+  bool reserve(unsigned int);
+
+  const size_t count();
+  void copy(Vec<RefCountCacheHashEntry *> &items);
+
+  typedef typename TSHashTable<RefCountCacheHashing>::iterator iterator_type;
+  typedef typename TSHashTable<RefCountCacheHashing>::self hash_type;
+  typedef typename TSHashTable<RefCountCacheHashing>::Location location_type;
+  TSHashTable<RefCountCacheHashing> *get_map();
+
+  Ptr<ProxyMutex> lock; // Lock
+
+private:
+  void metric_inc(RefCountCache_Stats metric_enum, int64_t data);
+
+  unsigned int part_num;
+
+  uint64_t max_size;
+  unsigned int max_items;
+
+  uint64_t size;
+  unsigned int items;
+
+  hash_type item_map;
+
+  PriorityQueue<RefCountCacheHashEntry *> expiry_queue;
+  RecRawStatBlock *rsb;
+};
+
+template <class C>
+RefCountCachePartition<C>::RefCountCachePartition(unsigned int part_num, uint64_t max_size, unsigned int max_items,
+                                                  RecRawStatBlock *rsb)
+{
+  this->part_num  = part_num;
+  this->max_size  = max_size;
+  this->max_items = max_items;
+  this->size      = 0;
+  this->items     = 0;
+  this->rsb       = rsb;
+
+  // Initialize lock
+  this->lock = new_ProxyMutex();
+}
+
+template <class C>
+Ptr<C>
+RefCountCachePartition<C>::get(uint64_t key)
+{
+  this->metric_inc(refcountcache_total_lookups_stat, 1);
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    // found
+    this->metric_inc(refcountcache_total_hits_stat, 1);
+    return make_ptr((C *)l.m_value->item.get());
+  } else {
+    return Ptr<C>();
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::put(uint64_t key, C *item, int size, int expire_time)
+{
+  this->metric_inc(refcountcache_total_inserts_stat, 1);
+  size += sizeof(C);
+  // Remove any colliding entries
+  this->erase(key);
+
+  // if we are full, and can't make space-- then don't store the item
+  if (this->is_full() && !this->reserve(size)) {
+    Debug("refcountcache", "partition %d is full-- not storing item key=%ld", this->part_num, key);
+    this->metric_inc(refcountcache_total_failed_inserts_stat, 1);
+    return;
+  }
+
+  // Create our value-- which has a ref to the `item`
+  RefCountCacheHashEntry *val = refCountCacheHashingValueAllocator.alloc();
+  val->set(item, key, size, expire_time);
+
+  // add expiry_entry to expiry queue, if the expire time is positive (otherwise it means don't expire)
+  if (expire_time >= 0) {
```

Comment:
I think I'd prefer to leave it as is, since it is only a check if an item has an expire time. having the value as a 0 check is also beneficial in the testing side of things :)